code/iaas/cluster-logic/src/main/java/io/cattle/iaas/cluster/process/ClusterUpdateConfig.java

```diff
@@ -0,0 +1,146 @@
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPostListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.json.JsonMapper;
+import io.cattle.platform.lock.LockCallbackNoReturn;
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.object.util.DataUtils;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+import io.cattle.platform.util.type.Priority;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import com.netflix.config.DynamicStringListProperty;
+import com.netflix.config.DynamicStringProperty;
+
+@Named
+public class ClusterUpdateConfig extends AbstractObjectProcessLogic implements ProcessPostListener, Priority {
+    static final DynamicStringListProperty CONFIG_ITEMS = ArchaiusUtil.getList("cluster.config.items");
+    static final DynamicStringProperty CLUSTER_SECUREDOCKER_PORT = ArchaiusUtil.getString("cluster.securedocker.port");
+
+    @Inject
+    ClusterHostMapDao clusterHostMapDao;
+
+    @Inject
+    ConfigItemStatusManager statusManager;
+
+    @Inject
+    ClusterManager clusterManager;
+
+    @Inject
+    JsonMapper jsonMapper;
+
+    @Inject
+    LockManager lockManager;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { "host.activate", "clusterhostmap.create", "clusterhostmap.remove" };
+    }
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        /** TODO: Depending on what discovery service is used to manage what hosts are in the cluster,
+         * this handler will have to do the appropriate thing (ie. make a call to swarm, make an HTTP call, update zookeeper, etc...)
+         * Actually, making the call to swarm should handle all these scenarios except for the node://
+         *
+         * Future: Deploy out socat process with server certs
+         */
+        Host cluster = null;
+        if (objectManager.isKind(state.getResource(), ClusterConstants.KIND)) {
+            cluster = (Host)state.getResource();
+        } else if (state.getResource() instanceof ClusterHostMap) {
+            Long clusterId = ((ClusterHostMap) state.getResource()).getClusterId();
+            cluster = objectManager.loadResource(Host.class, clusterId);
+        }
+        if (cluster == null) {
+            return null;
+        }
+        final Host finalCluster = cluster;
+        final ProcessState finalProcessState = state;
+        lockManager.lock(new ClusterLock(cluster), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                if (CommonStatesConstants.ACTIVE.equals(finalCluster.getState()) ||
+                        CommonStatesConstants.ACTIVATING.equals(finalCluster.getState())) {
+                    updateDiscoverySpec(finalCluster);
```

Comment:
We shouldn't load the discovery spec on every host add/activate for the cluster as this value is hardcoded anyway. Just do it once.